File: 0127.WordLadder.py
class Solution(object):
    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: Set[str]
        :rtype: int
        """
        queue = [(beginWord, 1)]
        visited = set()
        wordLen = len(beginWord)
        while queue:
            word, step = queue.pop(0)
            if word == endWord:
                return step
            for i in xrange(wordLen):
                for ch in 'abcdefghijklmnopqrstuvwxyz':
                    new = word[:i] + ch + word[i+1:]
                    if new in wordList and new not in visited:
                    #if (new in wordList or new == endWord) and new not in visited:
                        queue.append((new, step+1))
                        visited.add(new)
        return 0

File: 0127.WordLadder.go
func ladderLength(beginWord string, endWord string, wordList []string) int {
    Map := make(map[string]bool, len(wordList))
    for _, word := range wordList {
        Map[word] = true
    }
    queue := []string{beginWord}
    step := 1
    for len(queue) > 0 {
        nxt := []string{}
        for _, cur := range queue {
            if cur == endWord {
                return step
            }
            for i := 0; i < len(cur); i++ {
                for ch := 'a'; ch <= 'z'; ch++ {
                    tmp := cur[:i] + string(ch) + cur[i+1:]
                    if Map[tmp] {
                        nxt = append(nxt, tmp)
                        delete(Map, tmp)
                    }
                }
            }
        }
        step++
        queue = nxt
    }
    return 0
}